app/controllers/users_controller.rb

```diff
@@ -112,6 +111,24 @@ def show
     @phenotype_comment_replies.sort! { |b,a| a.created_at <=> b.created_at }
     @paginated_phenotype_replies = @phenotype_comment_replies
 
+    @last_30_papers = Set.new
+    if !@user.genotypes.nil?
+      # get the last 30 updated SNPs for this user
+      # we do it the other way around - check the last 20 updated papers for each type, see if their snps are in user's collection
+      # the upper if is important since without it even users without SNPs will trigger this
+      %w(Snpedia Mendeley GenomeGov Plos).each do |source|
+        "#{source}Paper".constantize.last(30).each do |p|
+          p.snps.each do |s|
+            @last_30_papers.add(p) if s.users.include? @user
```

Comment:
Try if `s.users.exists?(@user.id)` is faster. `s.users.include?(@user)` will probably load all the users, `exists?` runs the minimal query necessary to find out if the user can be found, without loading it.